### PR TITLE
Extended AD authenticate class

### PR DESCRIPTION
**Pull request by CeltonMcGrath**

### CODE REVIEW

**Comment by CeltonMcGrath** [comment]
Based off of LDAP Authenticate example.

File: src/tactic/active_directory/ad_authenticate.py [hunk 1]
```diff
@@ -94,7 +94,18 @@ def verify(my, login_name, password):
...
         if is_logged_in:
             my.load_user_data(base_login_name, domain)
-                
+        else:
+            # If AD authentication fails, attempt login via Tactic database
+            # (Only allow login for external users)
+            login = Login.get_by_login(base_login_name)
+            if login and login.get_value('location', no_exception=True) == 'external':
+                auth_class = "pyasm.security.TacticAuthenticate"
+                authenticate = Common.create_from_class_path(auth_class)
```
Comment (oboreo) [inline src/tactic/active_directory/ad_authenticate.py:103]: you need to import Common at Line 17

File: src/tactic/active_directory/ad_authenticate.py [hunk 2]
```diff
@@ -94,7 +94,18 @@ def verify(my, login_name, password):
...
+                is_authenticated = authenticate.verify(base_login_name, password)
+                if is_authenticated == True:
+                    my.internal = False
```
Comment (oboreo) [inline src/tactic/active_directory/ad_authenticate.py]: this instance variable is being used? If not, remove it

File: src/pyasm/search/upgrade/project/sthpw_upgrade.py [hunk 3]
```diff
@@ -21,8 +21,11 @@ class SthpwUpgrade(BaseUpgrade):
     #
     # 4.2.0.a01
     #
-
-
+    def upgrade_v4_2_0_a01_012(my):
+        my.run_Sql('''
```
Comment (oboreo) [inline src/pyasm/search/upgrade/project/sthpw_upgrade.py]: it's lowercase run_sql(
I suggest running python upgrade_db.py -f 
before submitting

**Pull request merged**